Fix annotation autorange on category axes

etpinard opened the pull request:
fixes https://github.com/plotly/plotly.js/issues/1768 by making the category axes `setConvert` methods a bit more of a mess. Note that annotation autorange uses [`ax.r2c`](https://github.com/plotly/plotly.js/blob/ffb3791989ff97de528fc66d0f9748fa2aedec43/src/components/annotations/calc_autorange.js#L61).

Oh well, with this patch here, it seems like all the cases are covered.

Together with https://github.com/plotly/plotly.js/pull/1748, we now have working category annotations on first draw and on pan/zoom :tada: :

![peek 2017-06-15 15-21](https://user-images.githubusercontent.com/6675409/27197965-63bcd19a-51de-11e7-862a-10e000385465.gif)

By the way @alexcjohnson I chose to add a new mock in https://github.com/plotly/plotly.js/commit/f5c038295616012143626272ee2bb7afdc520540 instead of appending `annotation-autorange.json` to test shapes on category axes as well. I hope you don't mind.

alexcjohnson commented:
Nice and simple, and this actually seems like it's the logical way for `ax.r2c` to work. No problem about splitting off the mock - you're right to include both x and y, shapes and annotations (what about images?) - though at some point I feel like a bit of consolidation may be in order.

But did that mock fail before the patch? The coordinates that lead to the shape and annotations bumping out the autorange didn't come from category names. For annotations we can use a category that's already in the data and let the text box or arrow add extra pixel padding, for shapes it seems like we may need to use a category that's not in the data, so `ax.categoryarray`?

etpinard commented:
@alexcjohnson 

> Nice and simple, and this actually seems like it's the logical way for ax.r2c to work.

Great.

> (what about images?)

Layout-image-induced autorange bumps aren't a thing ye (seet https://github.com/plotly/plotly.js/issues/1111) and the [`layout_image`](http://rickyreusser.com/plotly-mock-viewer/#layout_image) mock already has a category axis. But yeah, when #1111 will be completed, we should add an image or two to `category-autorange.json`.

> But did that mock fail before the patch?

Oops, I messed that up when I added arrows to the mock. Fixed in -> e4aa7a3deab5b26df802506dd1ea25610002a9d7

> for shapes it seems like we may need to use a category that's not in the data, so ax.categoryarray?

Good call here. That actually uncovered another bug. Done in -> aaf12ba9cbbb465a21d2680efafa43cf8822c3fb

Bonus: annotations and shapes with invalid category coordinates were broken too. Fixed in -> 0758b3b48b8591001c002ede5d1d5c4d8dcc65a5

alexcjohnson commented:
Beautiful. Gotta love it when all these extra edge cases pop out on fixing related issues.
💃
